LLM Observability docs (#22668)
DataDog/documentation: content/en/tracing/llm_observability/api.md

> ---

Author: I modeled this page as closely to Datadog's API docs as possible, but some formatting features are not available to me that are used there. I'm open to feedback on the way this is organized.

title: LLM Observability API
---
<div class="alert alert-warning">LLM Observability is in public beta, and this API is subject to change. If changes occur, Datadog will provide release notes with any applicable upgrade instructions.</a></div>

The LLM Observability API provides an interface for developers to send LLM-related traces and spans to Datadog. If your application is written in Python, you may prefer to use the [Python SDK][1].

## Spans API
Use this endpoint to send spans to Datadog. For details on the available kinds of spans, see [Span Kinds][2].

Endpoint
: `https://api.<DATADOG_SITE>/api/unstable/llm-obs/v1/trace/spans`

> <br />Replace `<DATADOG_SITE>` with the Datadog site you want to send data to (for example, `datadoghq.com` or `us3.datadoghq.com`).

Author: @alai97 The URL only actually works for two Datadog sites in my understanding, so autopopulating the URL could lead to confusing problems for users. I'll make sure this is still the case though!

Method
: `POST`

### Request

#### Headers (required)
- `DD-API-KEY=<YOUR_DATADOG_API_KEY>`
- `Content-Type="application/json"`

#### Body data (required)

{{< tabs >}}
{{% tab "Model" %}}
| Field | Type | Description                  |
|-------|------------------------------|------|
| data [*required*]|  [SpansRequestData](#spansrequestdata) | Entry point into the request body. |
{{% /tab %}}

{{% tab "Example" %}}
{{< code-block lang="json" >}}
{
  "data": {
    "type": "span",
    "attributes": {
      "ml_app": "weather-bot",
      "session_id": "1",
      "tags": [
        "env:dev",
        "service:weather-bot",

>         "env:staging",

Reviewer: Two `env` tags, not sure if that is intentional but just a note~

Author: @lievan, I should remove one of these `env` tags, correct?

Author: I've removed the `env:dev` tag, but let me know if it was supposed to be there.

| Field      | Type                          | Description                                |
|------------|-------------------------------|--------------------------------------------|
| type [*required*]        | string                        | Identifier for the request. Set to `span`. |
| attributes [*required*]  | [SpansPayload](#spanspayload) | The body of the request.  |

#### `SpansPayload`
| Field    | Type                | Description  |
|----------|---------------------|--------------|
| ml_app [*required*] | string              | The name of your LLM application.      |
| spans [*required*]  | [[Span](#span)] | A list of spans.           |
| tags                | [[Tag](#tag)]   | A list of top-level tags to apply to each span.        |
| session_id          | string              | The session the list of spans belongs to. Can be overridden or set on individual spans as well. |

#### `Tag`

Tags should be formatted as a list of strings (for example, `["user_handle:dog@gmail.com", "app_version:1.0.0"]`). They are meant to store contextual information surrounding the span.

## Eval metrics API
Use this endpoint to send evaluation metrics for a span to Datadog.

Endpoint

> : `https://api.<DATADOG_SITE>/api/unstable/llm-obs/v1/eval-metric`
> <br />Replace `<DATADOG_SITE>` with the Datadog site you want to send data to (for example, `datadoghq.com` or `us3.datadoghq.com`).

Reviewer: ```suggestion
: `https://api.{{< region-param key="dd_site" code="true" >}}/api/unstable/llm-obs/v1/eval-metric`
```

Method
: `POST`

Evaluation metrics require a `span_id` and `trace_id`. 

> - If you are not using the SDK, send the `span_id` and `trace_id` that you used to create your target span.
> - If you are using the SDK, obtain the `span_id` and `trace_id` by finding your target span, then accessing the `root_span.span_id` and the `root_span.trace_id` attributes.

Reviewer: ```suggestion
- If you are not using the LLM Observability SDK for Python, send the `span_id` and `trace_id` that you used to create your target span.
- If you are using the LLM Observability SDK for Python, obtain the `span_id` and `trace_id` by finding your target span, and accessing the span's `root_span.span_id` and the `root_span.trace_id` attributes.
```

|-------|------------------------------|------|
| data [*required*]  | [EvalMetricsRequestData](#evalmetricsrequestdata) | Entry point into the request body. |
{{% /tab %}}

{{% tab "Example" %}}
{{< code-block lang="json" >}}
{
  "data": {
    "type": "evaluation_metric",
    "attributes": {
      "metrics": [
        {
          "span_id": "61399242116139924211",
          "trace_id": "13932955089405749200",
          "timestamp": 1609459200,
          "metric_type": "categorical",
          "label": "Sentiment",
          "categorical_value": "Positive"
        },
        {
          "span_id": "20245611112024561111",
          "trace_id": "13932955089405749200",
          "metric_type": "score",
          "label": "Accuracy",
          "score_value": 3
        }
      ]
    }
  }
}
{{< /code-block >}}
{{% /tab %}}
{{< /tabs >}}

### Response

{{< tabs >}}
{{% tab "Model" %}}
| Field   | Type                        | Description                              | Guaranteed |
|---------|-----------------------------|------------------------------------------|------------|
| ID      | string                      | Response UUID generated upon submission. | Yes        |

> | metrics | [[EvalMetric](#evalmetric)] | A list of eval metrics.                  | Yes        |

Reviewer: ```suggestion
| metrics | [[EvalMetric](#evalmetric)] | A list of evaluation metrics.                  | Yes        |
```

{{% /tab %}}

{{% tab "Example" %}}
{{< code-block lang="json" >}}
{
  "data": {
    "type": "evaluation_metric",
    "id": "456f4567-e89b-12d3-a456-426655440000",
    "attributes": {
      "metrics": [
        {
          "id": "d4f36434-f0cd-47fc-884d-6996cee26da4",
          "span_id": "61399242116139924211",
          "trace_id": "13932955089405749200",
          "timestamp": 1609459200,
          "metric_type": "categorical",
          "label": "Sentiment",
          "categorical_value": "Positive"
        },
        {
          "id": "cdfc4fc7-e2f6-4149-9c35-edc4bbf7b525",
          "span_id": "20245611112024561111",
          "trace_id": "13932955089405749200",
          "metric_type": "score",
          "label": "Accuracy",
          "score_value": 3
        }
      ]
    }
  }
}
{{< /code-block >}}
{{% /tab %}}
{{< /tabs >}}

### API standards

#### Attribute

| Field   | Type         | Description                                         |
|---------|--------------|-----------------------------------------------------|

> | metrics [*required*] | [[EvalMetric](#evalmetric)] | A list of eval metrics for a given prompt-response pair. |

Reviewer: ```suggestion
| metrics [*required*] | [[EvalMetric](#evalmetric)] | A list of evaluation metrics for a given prompt-response pair. |
```

#### EvalMetric

| Field                  | Type   | Description  |
|------------------------|--------|--------------|
| ID                     | string | Evaluation metric UUID (generated upon submission). |

> | span_id [*required*]    | string | The ID of the span that this eval metric is associated with. This should be the span ID of the root span. |
> | trace_id [*required*]   | string | The ID of the trace that this eval metric is associated with. |

Reviewer: ```suggestion
| span_id [*required*]    | string | The ID of the span that this evaluation metric is associated with. This should be the span ID of the root span. |
| trace_id [*required*]   | string | The ID of the trace that this evaluation metric is associated with. |
```

| timestamp              | int64  | A UTC UNIX timestamp representing the time the request was sent. |
| metric_type [*required*]| string | The type of evaluation metric: `"categorical"` or `"score"`. |
| label [*required*]      | string | The unique name or label for the provided evaluation metric. |
| categorical_value [*required if the metric_type is "score"*]    | string | A string representing the category that the evaluation metric belongs to. |
| score_value [*required if the metric_type is "score"*]    | number | A score value of the evaluation metric. |
| flagged                | boolean| Flag content as inappropriate or incorrect. |
| annotation             | string | A generic string note about the provided evaluation metric. |

> #### EvalMetricsRequestData

Reviewer: +1 on hammering out the content in this section.

| Field      | Type            | Description  | 
|------------|-----------------|--------------|

> | type [*required*]      | "evaluation_metric" | Identifier for the request. Value depends on the API. |

Author: The type "evaluation_metric" seems strange, is there something else that should go here? "Value depends on the API" is also confusing, can you clarify?

Reviewer 2: So if someone is using the evaluation metrics endpoint they have to specify the `type` as `evaluation_metric`. It could be more clear if we directly tell them it has to be exactly `evaluation_metric`?

For "Value Depends on the API" -- agree it's confusing. It's supposed to communicate that the `type` depends on what endpoint you are using, e.g. the `POST /spans` the value would be `span`.

here's what we have for the `POST /spans` endpoint, should we also change this to be something similar?
<img width="713" alt="image" src="https://github.com/DataDog/documentation/assets/42917263/4453db83-9f2c-4eb4-b9b7-7ab0b32033be">

> | attributes [*required*] | [[EvalMetric](#evalmetric)] | The body of the request. Refer to one of the available API's attribute documentation. | 

Author: I found "Refer to one of the available API's attribute documenation" confusing here. Can you clarify what is meant by this?

Reviewer 2: I think we can just delete "Refer to one of the available API’s attribute documentation."

I think this documentation was written with the assumption that type/attributes could change for the same endpoint, but that's not the case here - type will always be `evaluation_metric`, and `attributes` will always be of type List[Evalmetric]

[1]: /tracing/llm_observability/sdk/

> [2]: /tracing/llm_observability/span_kinds/

Reviewer: ```suggestion
[2]: /tracing/llm_observability/span_kinds/
[3]: /getting_started/tagging/
```
